File: src/core/trace/reader.ts
/**
 * TraceReader — Query and read trace records from filesystem storage.
 *
 * Storage path: `<workspace>/.routa/traces/{day}/traces-{datetime}.jsonl`
 *
 * In serverless environments (Vercel), reads from /tmp/.routa/traces/
 * since that's where TraceWriter stores traces.
 *
 * Features:
 * - Filter traces by session, file, workspace, date range
 * - Retrieve individual traces by ID
 * - Export traces in standard Agent Trace JSON format
 * - Efficient file scanning with early termination on match
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";
import type { TraceRecord, TraceEventType } from "./types";

/**
 * Check if running in a serverless environment (e.g., Vercel)
 */
function isServerlessEnvironment(): boolean {
  return !!(process.env.VERCEL || process.env.AWS_LAMBDA_FUNCTION_NAME);
}

/**
 * Query parameters for filtering traces.
 */
export interface TraceQuery {
  /** Filter by session ID */
  sessionId?: string;
  /** Filter by workspace ID */
  workspaceId?: string;
  /** Filter by file path */
  file?: string;
  /** Filter by event type */
  eventType?: TraceEventType;
  /** Start date (YYYY-MM-DD or ISO 8601) */
  startDate?: string;
  /** End date (YYYY-MM-DD or ISO 8601) */
  endDate?: string;
  /** Maximum number of traces to return */
  limit?: number;
  /** Skip N traces (for pagination) */
  offset?: number;
}

/**
 * Trace statistics for a workspace.
 */
export interface TraceStats {
  totalDays: number;
  totalFiles: number;
  totalRecords: number;
  uniqueSessions: number;
  eventTypes: Record<string, number>;
}

/**
 * TraceReader provides querying capabilities over stored traces.
 *
 * In serverless environments, reads from /tmp/.routa/traces/ since
 * that's the only writable location.
 */
export class TraceReader {
  /** Base directory for trace files (e.g., "/project/.routa/traces") */
  readonly #baseDir: string;

  /**
   * Create a new TraceReader with the given workspace root.
   *
   * Traces are read from `<workspace_root>/.routa/traces/`.
   * In serverless environments, reads from `/tmp/.routa/traces/`.
   */
  constructor(workspaceRoot: string) {
    const basePath = isServerlessEnvironment() ? "/tmp" : workspaceRoot;
    this.#baseDir = path.join(basePath, ".routa", "traces");
  }

  /**
   * Create a TraceReader with a custom base directory.
   */
  static withBaseDir(baseDir: string): TraceReader {
    return new TraceReader(baseDir.replace(/\.routa\/traces$/, ""));
  }

  /**
   * Query traces based on the provided filter parameters.
   *
   * Returns traces sorted by timestamp (newest first).
   */
  async query(query: TraceQuery = {}): Promise<TraceRecord[]> {
    // If traces directory doesn't exist, return empty result
    try {
      await fs.access(this.#baseDir);
    } catch {
      return [];
    }

    const traces: TraceRecord[] = [];

    // Get all day directories
    const dayDirs = await this.#listDayDirs();

    // Apply date filtering if specified
    const filteredDays = this.#filterDaysByDate(dayDirs, query);

    // Read traces from each day directory
    for (const dayDir of filteredDays) {
      const traceFiles = await this.#listTraceFiles(dayDir);

      for (const traceFile of traceFiles) {
        const content = await fs.readFile(traceFile, "utf-8");

        for (const line of content.split("\n").filter(Boolean)) {
          try {
            const record: TraceRecord = JSON.parse(line);
            if (this.#matchesQuery(record, query)) {
              traces.push(record);
            }
          } catch {
            // Skip invalid lines
          }
        }

        // Early termination if we have enough results
        const needed = (query.limit ?? Infinity) + (query.offset ?? 0);
        if (traces.length >= needed) {
          break;
        }
      }
    }

    // Sort by timestamp (oldest first for chronological reading) and apply pagination
    traces.sort((a, b) => a.timestamp.localeCompare(b.timestamp));

    const offset = query.offset ?? 0;
    const limit = query.limit ?? traces.length;

    return traces.slice(offset, offset + limit);
  }

  /**
   * Get a single trace by its ID.
   */
  async getById(id: string): Promise<TraceRecord | null> {
    try {
      await fs.access(this.#baseDir);
    } catch {
      return null;
    }

    const dayDirs = await this.#listDayDirs();

    for (const dayDir of dayDirs) {
      const traceFiles = await this.#listTraceFiles(dayDir);

      for (const traceFile of traceFiles) {
        const content = await fs.readFile(traceFile, "utf-8");

        for (const line of content.split("\n").filter(Boolean)) {
          try {
            const record: TraceRecord = JSON.parse(line);
            if (record.id === id) {
              return record;
            }
          } catch {
            // Skip invalid lines
          }
        }
      }
    }

    return null;
  }

  /**
   * Export traces matching the query in Agent Trace JSON format.
   *
   * Returns a JSON array of trace records.
   */
  async export(query: TraceQuery = {}): Promise<unknown> {
    const traces = await this.query(query);
    return traces;
  }

  /**
   * Get trace statistics for a workspace.
   */
  async stats(): Promise<TraceStats> {
    const defaultStats: TraceStats = {
      totalDays: 0,
      totalFiles: 0,
      totalRecords: 0,
      uniqueSessions: 0,
      eventTypes: {},
    };

    try {
      await fs.access(this.#baseDir);
    } catch {
      return defaultStats;
    }

    const stats: TraceStats = { ...defaultStats, eventTypes: {} };
    const sessions = new Set<string>();

    const dayDirs = await this.#listDayDirs();
    stats.totalDays = dayDirs.length;

    for (const dayDir of dayDirs) {
      const traceFiles = await this.#listTraceFiles(dayDir);
      stats.totalFiles += traceFiles.length;

      for (const traceFile of traceFiles) {
        const content = await fs.readFile(traceFile, "utf-8");
        const lines = content.split("\n").filter(Boolean);
        stats.totalRecords += lines.length;

        for (const line of lines) {
          try {
            const record: TraceRecord = JSON.parse(line);
            sessions.add(record.sessionId);
            stats.eventTypes[record.eventType] =
              (stats.eventTypes[record.eventType] ?? 0) + 1;
          } catch {
            // Skip invalid lines
          }
        }
      }
    }

    stats.uniqueSessions = sessions.size;
    return stats;
  }

  /**
   * List all day directories sorted newest first.
   */
  async #listDayDirs(): Promise<string[]> {
    const entries = await fs.readdir(this.#baseDir, { withFileTypes: true });
    const dirs = entries
      .filter((e) => e.isDirectory())
      .map((e) => path.join(this.#baseDir, e.name));
    return dirs.sort().reverse();
  }

  /**
   * List all trace files in a day directory sorted by name.
   */
  async #listTraceFiles(dayDir: string): Promise<string[]> {
    const entries = await fs.readdir(dayDir, { withFileTypes: true });
    const files = entries
      .filter((e) => e.isFile() && e.name.endsWith(".jsonl"))
      .map((e) => path.join(dayDir, e.name));
    return files.sort().reverse();
  }

  /**
   * Filter day directories by date range.
   */
  #filterDaysByDate(dayDirs: string[], query: TraceQuery): string[] {
    const filtered: string[] = [];

    for (const dayDir of dayDirs) {
      const dayName = path.basename(dayDir);

      if (!this.#isValidDateFormat(dayName)) {
        continue;
      }

      if (query.startDate && dayName < query.startDate) {
        continue;
      }

      if (query.endDate && dayName > query.endDate) {
        continue;
      }

      filtered.push(dayDir);
    }

    return filtered;
  }

  /**
   * Check if a date string is valid YYYY-MM-DD format.
   */
  #isValidDateFormat(dateStr: string): boolean {
    return /^\d{4}-\d{2}-\d{2}$/.test(dateStr);
  }

  /**
   * Check if a trace record matches the query parameters.
   */
  #matchesQuery(record: TraceRecord, query: TraceQuery): boolean {
    if (query.sessionId && record.sessionId !== query.sessionId) {
      return false;
    }

    if (query.workspaceId && record.workspaceId !== query.workspaceId) {
      return false;
    }

    if (query.file) {
      const fileMatches = record.files?.some((f) => f.path === query.file) ?? false;
      if (!fileMatches) {
        return false;
      }
    }

    if (query.eventType && record.eventType !== query.eventType) {
      return false;
    }

    return true;
  }
}

// ─── Singleton Instance ───────────────────────────────────────────────────

const GLOBAL_KEY = "__trace_readers__";

type TraceReaderCache = Map<string, TraceReader>;

function getReaderCache(): TraceReaderCache {
  const g = globalThis as Record<string, unknown>;
  if (!g[GLOBAL_KEY]) {
    g[GLOBAL_KEY] = new Map<string, TraceReader>();
  }
  return g[GLOBAL_KEY] as TraceReaderCache;
}

/**
 * Get or create a TraceReader for the given cwd.
 */
export function getTraceReader(cwd: string): TraceReader {
  const cache = getReaderCache();
  let reader = cache.get(cwd);
  if (!reader) {
    reader = new TraceReader(cwd);
    cache.set(cwd, reader);
  }
  return reader;
}

File: crates/routa-core/src/trace/reader.rs
//! TraceReader — Query and read trace records from filesystem storage.
//!
//! Storage path: `<workspace>/.routa/traces/{day}/traces-{datetime}.jsonl`
//!
//! Features:
//! - Filter traces by session, file, workspace, date range
//! - Retrieve individual traces by ID
//! - Export traces in standard Agent Trace JSON format
//! - Efficient file scanning with early termination on match

use std::path::{Path, PathBuf};
use std::collections::HashMap;
use serde_json::Value;

use super::types::TraceRecord;

/// Query parameters for filtering traces.
#[derive(Debug, Clone, Default)]
pub struct TraceQuery {
    /// Filter by session ID
    pub session_id: Option<String>,
    /// Filter by workspace ID
    pub workspace_id: Option<String>,
    /// Filter by file path
    pub file: Option<String>,
    /// Filter by event type
    pub event_type: Option<String>,
    /// Start date (ISO 8601 or YYYY-MM-DD)
    pub start_date: Option<String>,
    /// End date (ISO 8601 or YYYY-MM-DD)
    pub end_date: Option<String>,
    /// Maximum number of traces to return
    pub limit: Option<usize>,
    /// Skip N traces (for pagination)
    pub offset: Option<usize>,
}

/// TraceReader provides querying capabilities over stored traces.
#[derive(Clone)]
pub struct TraceReader {
    /// Base directory for trace files (e.g., "/project/.routa/traces")
    base_dir: PathBuf,
}

impl TraceReader {
    /// Create a new TraceReader with the given workspace root.
    ///
    /// Traces are read from `<workspace_root>/.routa/traces/`.
    pub fn new(workspace_root: impl AsRef<Path>) -> Self {
        let base_dir = workspace_root.as_ref().join(".routa").join("traces");
        Self {
            base_dir,
        }
    }

    /// Create a TraceReader with a custom base directory.
    pub fn with_base_dir(base_dir: impl AsRef<Path>) -> Self {
        Self {
            base_dir: base_dir.as_ref().to_path_buf(),
        }
    }

    /// Query traces based on the provided filter parameters.
    ///
    /// Returns traces sorted by timestamp (newest first).
    pub async fn query(&self, query: &TraceQuery) -> Result<Vec<TraceRecord>, TraceReadError> {
        // If traces directory doesn't exist, return empty result
        if !self.base_dir.exists() {
            return Ok(Vec::new());
        }

        let mut traces = Vec::new();

        // Get all day directories
        let mut day_dirs = collect_dirs(&self.base_dir).await?;

        // Sort day directories (newest first)
        day_dirs.sort_by(|a, b| b.cmp(a));

        // Apply date filtering if specified
        let filtered_days = if let (Some(start), Some(end)) = (&query.start_date, &query.end_date) {
            self.filter_days_by_range(&day_dirs, start, end)?
        } else if let Some(start) = &query.start_date {
            self.filter_days_since(&day_dirs, start)?
        } else if let Some(end) = &query.end_date {
            self.filter_days_until(&day_dirs, end)?
        } else {
            day_dirs
        };

        // Read traces from each day directory
        for day_dir in filtered_days {
            // Read all JSONL files in the day directory
            let mut trace_files = collect_jsonl_files(&day_dir).await?;

            // Sort trace files by name (which contains timestamp)
            trace_files.sort_by(|a, b| b.cmp(a));

            for trace_file in trace_files {
                let content = tokio::fs::read_to_string(&trace_file).await
                    .map_err(|e| TraceReadError::Io(format!("Failed to read trace file: {}", e)))?;

                for line in content.lines() {
                    if let Ok(record) = serde_json::from_str::<TraceRecord>(line) {
                        if self.matches_query(&record, query) {
                            traces.push(record);
                        }
                    }
                }

                // Early termination if we have enough results
                if let (Some(limit), Some(offset)) = (query.limit, query.offset) {
                    if traces.len() >= limit + offset {
                        break;
                    }
                } else if let Some(limit) = query.limit {
                    if traces.len() >= limit {
                        break;
                    }
                }
            }
        }

        // Sort by timestamp (newest first) and apply pagination
        traces.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));

        let offset = query.offset.unwrap_or(0);
        let limit = query.limit.unwrap_or(traces.len());

        Ok(traces.into_iter().skip(offset).take(limit).collect())
    }

    /// Get a single trace by its ID.
    pub async fn get_by_id(&self, id: &str) -> Result<Option<TraceRecord>, TraceReadError> {
        if !self.base_dir.exists() {
            return Ok(None);
        }

        // Search through all trace files
        let day_dirs = collect_dirs(&self.base_dir).await?;

        for day_dir in day_dirs {
            let trace_files = collect_jsonl_files(&day_dir).await?;

            for trace_file in trace_files {
                let content = tokio::fs::read_to_string(&trace_file).await
                    .map_err(|e| TraceReadError::Io(format!("Failed to read trace file: {}", e)))?;

                for line in content.lines() {
                    if let Ok(record) = serde_json::from_str::<TraceRecord>(line) {
                        if record.id == id {
                            return Ok(Some(record));
                        }
                    }
                }
            }
        }

        Ok(None)
    }

    /// Export traces matching the query in Agent Trace JSON format.
    ///
    /// Returns a JSON array of trace records.
    pub async fn export(&self, query: &TraceQuery) -> Result<Value, TraceReadError> {
        let traces = self.query(query).await?;
        let traces_json: Value = serde_json::to_value(traces)
            .map_err(|e| TraceReadError::Serialization(format!("Failed to serialize traces: {}", e)))?;
        Ok(traces_json)
    }

    /// Get trace statistics for a workspace.
    pub async fn stats(&self) -> Result<TraceStats, TraceReadError> {
        if !self.base_dir.exists() {
            return Ok(TraceStats::default());
        }

        let mut stats = TraceStats::default();

        let day_dirs = collect_dirs(&self.base_dir).await?;

        for day_dir in day_dirs {
            stats.total_days += 1;

            let trace_files = collect_jsonl_files(&day_dir).await?;
            stats.total_files += trace_files.len() as u32;

            for trace_file in trace_files {
                let content = tokio::fs::read_to_string(&trace_file).await
                    .map_err(|e| TraceReadError::Io(format!("Failed to read trace file: {}", e)))?;

                stats.total_records += content.lines().count();

                // Track sessions and event types
                for line in content.lines() {
                    if let Ok(record) = serde_json::from_str::<TraceRecord>(line) {
                        stats.sessions.insert(record.session_id.clone());
                        let event_type_str = format!("{:?}", record.event_type);
                        *stats.event_types.entry(event_type_str).or_insert(0) += 1;
                    }
                }
            }
        }

        stats.unique_sessions = stats.sessions.len() as u32;

        Ok(stats)
    }

    /// Check if a trace record matches the query parameters.
    fn matches_query(&self, record: &TraceRecord, query: &TraceQuery) -> bool {
        if let Some(ref session_id) = query.session_id {
            if &record.session_id != session_id {
                return false;
            }
        }

        if let Some(ref workspace_id) = query.workspace_id {
            if record.workspace_id.as_ref() != Some(workspace_id) {
                return false;
            }
        }

        if let Some(ref file) = query.file {
            let file_matches = record.files.iter().any(|f| &f.path == file);
            if !file_matches {
                return false;
            }
        }

        if let Some(ref event_type) = query.event_type {
            let record_type = format!("{:?}", record.event_type).to_lowercase();
            let query_lower = event_type.to_lowercase();
            if record_type != query_lower {
                // Also check snake_case variant
                let record_type_snake = to_snake_case(&format!("{:?}", record.event_type));
                if record_type_snake != query_lower {
                    return false;
                }
            }
        }

        true
    }

    /// Filter day directories by date range.
    fn filter_days_by_range(
        &self,
        day_dirs: &[PathBuf],
        start: &str,
        end: &str,
    ) -> Result<Vec<PathBuf>, TraceReadError> {
        let start_date = self.parse_date(start)?;
        let end_date = self.parse_date(end)?;

        Ok(day_dirs
            .iter()
            .filter(|path| {
                if let Some(date_str) = path.file_name().and_then(|n| n.to_str()) {
                    if let Ok(date) = self.parse_date(date_str) {
                        return date >= start_date && date <= end_date;
                    }
                }
                false
            })
            .cloned()
            .collect())
    }

    /// Filter day directories since a start date.
    fn filter_days_since(
        &self,
        day_dirs: &[PathBuf],
        start: &str,
    ) -> Result<Vec<PathBuf>, TraceReadError> {
        let start_date = self.parse_date(start)?;

        Ok(day_dirs
            .iter()
            .filter(|path| {
                if let Some(date_str) = path.file_name().and_then(|n| n.to_str()) {
                    if let Ok(date) = self.parse_date(date_str) {
                        return date >= start_date;
                    }
                }
                false
            })
            .cloned()
            .collect())
    }

    /// Filter day directories until an end date.
    fn filter_days_until(
        &self,
        day_dirs: &[PathBuf],
        end: &str,
    ) -> Result<Vec<PathBuf>, TraceReadError> {
        let end_date = self.parse_date(end)?;

        Ok(day_dirs
            .iter()
            .filter(|path| {
                if let Some(date_str) = path.file_name().and_then(|n| n.to_str()) {
                    if let Ok(date) = self.parse_date(date_str) {
                        return date <= end_date;
                    }
                }
                false
            })
            .cloned()
            .collect())
    }

    /// Parse a date string (YYYY-MM-DD or ISO 8601).
    fn parse_date(&self, date_str: &str) -> Result<chrono::NaiveDate, TraceReadError> {
        let trimmed = date_str.split('T').next().unwrap_or(date_str);
        chrono::NaiveDate::parse_from_str(trimmed, "%Y-%m-%d")
            .map_err(|e| TraceReadError::InvalidDate(format!("Invalid date '{}': {}", date_str, e)))
    }
}

/// Helper function to collect directories from a path.
async fn collect_dirs(path: &Path) -> Result<Vec<PathBuf>, TraceReadError> {
    let mut dirs = Vec::new();
    let mut readdir = tokio::fs::read_dir(path)
        .await
        .map_err(|e| TraceReadError::Io(format!("Failed to read directory: {}", e)))?;

    while let Some(entry) = readdir.next_entry().await
        .map_err(|e| TraceReadError::Io(format!("Failed to read dir entry: {}", e)))?
    {
        let path = entry.path();
        if path.is_dir() {
            dirs.push(path);
        }
    }

    Ok(dirs)
}

/// Helper function to collect JSONL files from a path.
async fn collect_jsonl_files(path: &Path) -> Result<Vec<PathBuf>, TraceReadError> {
    let mut files = Vec::new();
    let mut readdir = tokio::fs::read_dir(path)
        .await
        .map_err(|e| TraceReadError::Io(format!("Failed to read directory: {}", e)))?;

    while let Some(entry) = readdir.next_entry().await
        .map_err(|e| TraceReadError::Io(format!("Failed to read dir entry: {}", e)))?
    {
        let path = entry.path();
        if path.is_file() {
            if let Some(ext) = path.extension() {
                if ext == "jsonl" {
                    files.push(path);
                }
            }
        }
    }

    Ok(files)
}

/// Convert a string to snake_case.
fn to_snake_case(s: &str) -> String {
    s.chars()
        .enumerate()
        .map(|(i, c)| {
            if c.is_uppercase() {
                if i > 0 {
                    format!("_{}", c.to_lowercase().collect::<String>())
                } else {
                    c.to_lowercase().collect::<String>()
                }
            } else {
                c.to_string()
            }
        })
        .collect()
}

/// Statistics about stored traces.
#[derive(Debug, Clone, Default, serde::Serialize)]
pub struct TraceStats {
    pub total_days: u32,
    pub total_files: u32,
    pub total_records: usize,
    pub unique_sessions: u32,
    #[serde(skip)]
    pub sessions: std::collections::HashSet<String>,
    pub event_types: HashMap<String, u32>,
}

/// Error type for trace reading operations.
#[derive(Debug, thiserror::Error)]
pub enum TraceReadError {
    #[error("IO error: {0}")]
    Io(String),

    #[error("Serialization error: {0}")]
    Serialization(String),

    #[error("Invalid date: {0}")]
    InvalidDate(String),
}
